src/analysis/natural_parameterization.lean

```diff
@@ -0,0 +1,229 @@
+  rw evariation_on.subsingleton f (λ y hy z hz, hs hy.1 hz.1 : (s ∩ Icc x x).subsingleton),
+  simp only [sub_self, mul_zero, ennreal.of_real_zero],
+end
+
+-- Could do without `hl` but it means we have to check whether `s` is subsingleton
+-- and if not deduce that `hl` necessarily holds.
+lemma is_linearly_parameterized_on_by.iff_ordered (f : ℝ → E) (s : set ℝ) {l : ℝ} (hl : 0 ≤ l):
+  is_linearly_parameterized_on_by f s l ↔
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), (x ≤ y) →
+    evariation_on f (s ∩ Icc x y) = ennreal.of_real (l * (y - x)) :=
+begin
+  refine ⟨λ h x xs y ys xy, h xs ys, λ h x xs y ys, _⟩,
+  rcases le_total x y with xy|yx,
+  { exact h xs ys xy, },
+  { rw [evariation_on.subsingleton, ennreal.of_real_of_nonpos],
+    { exact mul_nonpos_of_nonneg_of_nonpos hl (sub_nonpos_of_le yx), },
+    { rintro z ⟨zs, xz, zy⟩ w ⟨ws, xw, wy⟩,
+      cases le_antisymm (zy.trans yx) xz,
+      cases le_antisymm (wy.trans yx) xw,
+      refl, }, },
+end
+
+lemma is_linearly_parameterized_on_by.iff_variation_on_from_to_eq
+  (f : ℝ → E) (s : set ℝ) {l : ℝ} (hl : 0 ≤ l) :
+  is_linearly_parameterized_on_by f s l ↔ (has_locally_bounded_variation_on f s ∧
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), variation_on_from_to f s x y = l * (y - x)) :=
+begin
+  split,
+  { rintro h, refine ⟨h.has_locally_bounded_variation_on, λ x xs y ys, _⟩,
+    rw is_linearly_parameterized_on_by.iff_ordered f s hl at h,
+    rcases le_total x y with xy|yx,
+    { rw [variation_on_from_to.eq_of_le f s xy, h xs ys xy,
+          ennreal.to_real_of_real (mul_nonneg hl (sub_nonneg.mpr xy))], },
+    { rw [variation_on_from_to.eq_of_ge f s yx, h ys xs yx ,
+          ennreal.to_real_of_real  (mul_nonneg hl (sub_nonneg.mpr yx)),
+          mul_comm l, mul_comm l, ←neg_mul, neg_sub], }, },
+  { rw is_linearly_parameterized_on_by.iff_ordered f s hl,
+    rintro h x xs y ys xy,
+    rw [←h.2 xs ys, variation_on_from_to.eq_of_le f s xy,
+        ennreal.of_real_to_real (h.1 x y xs ys)], },
+end
+
+lemma is_linearly_parameterized_on_by_zero_iff (f : ℝ → E) (s : set ℝ) :
+  is_linearly_parameterized_on_by f s 0 ↔ ∀ x y ∈ s, edist (f x) (f y) = 0 :=
+begin
+  dsimp [is_linearly_parameterized_on_by],
+  simp only [zero_mul, ennreal.of_real_zero, ←evariation_on.eq_zero_iff],
+  split,
+  { by_contra',
+    obtain ⟨h, hfs⟩ := this,
+    simp_rw evariation_on.eq_zero_iff at hfs h,
+    push_neg at hfs,
+    obtain ⟨x, xs, y, ys, hxy⟩ := hfs,
+    rcases le_total x y with xy|yx,
+    { exact hxy (h xs ys x ⟨xs, le_rfl, xy⟩ y ⟨ys, xy, le_rfl⟩), },
+    { rw edist_comm at hxy,
+      exact hxy (h ys xs y ⟨ys, le_rfl, yx⟩ x ⟨xs, yx, le_rfl⟩), }, },
+  { rintro h x xs y ys,
+    refine le_antisymm _ (zero_le'),
+    rw ←h,
+    exact evariation_on.mono f (inter_subset_left s (Icc x y)), },
+end
+
+lemma is_linearly_parameterized_on_by.ratio {f : ℝ → E} {s t : set ℝ} {φ : ℝ → ℝ}
+  (φm : monotone_on φ s) (φst : s.maps_to φ t) (φst' : s.surj_on φ t)
+  {l l' : ℝ} (hl : 0 ≤ l) (hl' : 0 < l')
```

Comment:
again, `hl` and `hl'` are probably useless here.

Reply:
I don't think they are, but anyway, with the `real -> nnreal` transition, only `hl'` remains